File: custom-logs/send_logs.py
"""Send custom OTLP logs to the ApexData collector.

Uses the OpenTelemetry LoggingHandler: standard `logging` calls are shipped to the
collector as OTLP LogRecords. Each request is handled inside a trace span, so the
log records carry TraceId/SpanId.

Usage:
    python send_logs.py

Environment variables:
    OTEL_EXPORTER_OTLP_ENDPOINT  - Collector endpoint (default: localhost:4317)
    OTEL_EXPORTER_OTLP_HEADERS   - Auth headers (default: none)
    OTEL_SERVICE_NAME            - Service name (default: custom-logs-python)
    OTEL_EXPORTER_OTLP_INSECURE  - "true" disables TLS (default: false)
"""

import logging
import os
import random
import signal
import socket
import time

from opentelemetry._logs import set_logger_provider
from opentelemetry.exporter.otlp.proto.grpc._log_exporter import OTLPLogExporter
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.trace import get_tracer, set_tracer_provider


def get_env(key, fallback):
    return os.environ.get(key) or fallback


def parse_headers(headers_str):
    headers = {}
    for pair in headers_str.split(","):
        pair = pair.strip()
        if "=" in pair:
            key, _, value = pair.partition("=")
            # gRPC metadata keys must be lowercase; normalize here so that
            # env vars like "Authorization=…" work without manual casing.
            key, value = key.strip().lower(), value.strip()
            if key and value:
                headers[key] = value
    return headers


def main():
    endpoint = get_env("OTEL_EXPORTER_OTLP_ENDPOINT", "localhost:4317")
    headers_str = get_env("OTEL_EXPORTER_OTLP_HEADERS", "")
    service_name = get_env("OTEL_SERVICE_NAME", "custom-logs-python")
    insecure = get_env("OTEL_EXPORTER_OTLP_INSECURE", "false") == "true"
    headers = parse_headers(headers_str)

    # The gRPC exporter wants host:port — strip any scheme.
    endpoint = endpoint.replace("https://", "").replace("http://", "")

    print("=== Custom Logs Sender (Python/gRPC) ===")
    print(f"Endpoint: {endpoint}")
    print(f"Service:  {service_name}")
    print(f"Headers:  {len(headers)} configured\n")

    resource = Resource.create({
        "service.name": service_name,
        "service.version": "1.0.0",
        "host.name": socket.gethostname(),
        "deployment.environment": "demo",
    })

    logger_provider = LoggerProvider(resource=resource)
    set_logger_provider(logger_provider)
    exporter = OTLPLogExporter(endpoint=endpoint, headers=headers, insecure=insecure)
    logger_provider.add_log_record_processor(BatchLogRecordProcessor(exporter))

    handler = LoggingHandler(level=logging.NOTSET, logger_provider=logger_provider)
    # OTel's LoggingHandler formats the record into the OTLP Body. Use a bare
    # "%(message)s" formatter so the Body is just the message — the level travels
    # in SeverityText, not baked into the text. (basicConfig would otherwise apply
    # logging.BASIC_FORMAT here.)
    handler.setFormatter(logging.Formatter("%(message)s"))
    logging.basicConfig(level=logging.INFO, handlers=[handler])
    log = logging.getLogger("custom-logs")

    tracer_provider = TracerProvider(resource=resource)
    set_tracer_provider(tracer_provider)
    tracer = get_tracer("custom-logs")

    running = {"on": True}

    def stop(*_):
        running["on"] = False

    signal.signal(signal.SIGINT, stop)
    signal.signal(signal.SIGTERM, stop)

    print("Sending logs every 2 seconds... (Ctrl+C to stop)")
    print("Severities: INFO (normal), WARN (slow), ERROR (failed)\n")

    routes = ["/api/users", "/api/orders", "/api/checkout", "/healthz"]
    iteration = 0
    while running["on"]:
        iteration += 1
        route = random.choice(routes)
        duration_ms = random.randint(0, 500)
        status = 200

        # Each request handled inside a span -> logs carry TraceId/SpanId.
        with tracer.start_as_current_span("handle_request"):
            roll = random.random()
            if roll < 0.1:
                status = 500
                log.error("request failed", extra={
                    "http.method": "GET",
                    "http.route": route,
                    "http.status_code": status,
                    "duration_ms": duration_ms,
                    "error": "upstream timeout",
                })
            elif duration_ms > 350:
                log.warning("slow request", extra={
                    "http.method": "GET",
                    "http.route": route,
                    "http.status_code": status,
                    "duration_ms": duration_ms,
                })
            else:
                log.info("request handled", extra={
                    "http.method": "GET",
                    "http.route": route,
                    "http.status_code": status,
                    "duration_ms": duration_ms,
                })

        if iteration % 10 == 0:
            print(f"  [{iteration}] logs sent")
        time.sleep(2)

    print(f"\nStopping after {iteration} iterations...")
    logger_provider.shutdown()
    tracer_provider.shutdown()
    print("Done!")


if __name__ == "__main__":
    main()

File: custom-logs/send_logs.js
// Send custom OTLP logs to the ApexData collector (JavaScript example).
//
// Uses the raw @opentelemetry/sdk-logs API: logger.emit() ships records as OTLP
// LogRecords. Each request is emitted inside a trace span, so records carry
// TraceId/SpanId. Apps already using Winston can instead bridge via
// @opentelemetry/winston-transport (see README).
//
// Usage:
//   npm install && npm start
//
// Environment variables:
//   OTEL_EXPORTER_OTLP_ENDPOINT  - Collector endpoint (default: localhost:4317)
//   OTEL_EXPORTER_OTLP_HEADERS   - Auth headers (default: none)
//   OTEL_SERVICE_NAME            - Service name (default: custom-logs-js-demo)
//   OTEL_EXPORTER_OTLP_INSECURE  - "true" disables TLS (default: false)

'use strict';

const os = require('os');
const grpc = require('@grpc/grpc-js');
const { context, trace, propagation } = require('@opentelemetry/api');
const { AsyncLocalStorageContextManager } = require('@opentelemetry/context-async-hooks');
const { SeverityNumber } = require('@opentelemetry/api-logs');
const { LoggerProvider, BatchLogRecordProcessor } = require('@opentelemetry/sdk-logs');
const { OTLPLogExporter } = require('@opentelemetry/exporter-logs-otlp-grpc');
const { resourceFromAttributes } = require('@opentelemetry/resources');
const { BasicTracerProvider, AlwaysOnSampler } = require('@opentelemetry/sdk-trace-base');
const {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} = require('@opentelemetry/semantic-conventions');

// Register the async-local-storage context manager so that context.with() propagates
// span context to log records (TraceId/SpanId correlation).
context.setGlobalContextManager(new AsyncLocalStorageContextManager());

function getEnv(key, fallback) {
  const v = process.env[key];
  return v !== undefined && v !== '' ? v : fallback;
}

function buildMetadata(headersStr) {
  const metadata = new grpc.Metadata();
  for (const pair of headersStr.split(',')) {
    const trimmed = pair.trim();
    const idx = trimmed.indexOf('=');
    if (idx > 0) {
      // gRPC metadata keys must be lowercase.
      const key = trimmed.slice(0, idx).trim().toLowerCase();
      const value = trimmed.slice(idx + 1).trim();
      if (key && value) metadata.set(key, value);
    }
  }
  return metadata;
}

function main() {
  let endpoint = getEnv('OTEL_EXPORTER_OTLP_ENDPOINT', 'localhost:4317');
  const headersStr = getEnv('OTEL_EXPORTER_OTLP_HEADERS', '');
  const serviceName = getEnv('OTEL_SERVICE_NAME', 'custom-logs-js-demo');
  const insecure = getEnv('OTEL_EXPORTER_OTLP_INSECURE', 'false') === 'true';

  // The OTLP gRPC exporter wants a bare host:port — strip any scheme.
  endpoint = endpoint.replace(/^https?:\/\//, '');

  const metadata = buildMetadata(headersStr);

  console.log('=== Custom Logs Sender (JS/gRPC) ===');
  console.log(`Endpoint: ${endpoint}`);
  console.log(`Service:  ${serviceName}`);
  console.log(`Headers:  ${Object.keys(metadata.toJSON()).length} configured\n`);

  const resource = resourceFromAttributes({
    [ATTR_SERVICE_NAME]: serviceName,
    [ATTR_SERVICE_VERSION]: '1.0.0',
    'host.name': os.hostname(),
    'deployment.environment': 'demo',
  });

  const exporter = new OTLPLogExporter({
    url: endpoint,
    metadata,
    credentials: insecure
      ? grpc.credentials.createInsecure()
      : grpc.credentials.createSsl(),
  });

  const loggerProvider = new LoggerProvider({
    resource,
    processors: [new BatchLogRecordProcessor(exporter)],
  });

  // Trace provider only mints span contexts so log records carry TraceId/SpanId.
  // Spans themselves are not exported (no span processor) — that is intentional.
  const tracerProvider = new BasicTracerProvider({
    resource,
    sampler: new AlwaysOnSampler(),
  });
  const tracer = tracerProvider.getTracer('custom-logs');
  const logger = loggerProvider.getLogger('custom-logs', '1.0.0');

  console.log('Sending logs every 2 seconds... (Ctrl+C to stop)');
  console.log('Severities: INFO (normal), WARN (slow), ERROR (failed)\n');

  const routes = ['/api/users', '/api/orders', '/api/checkout', '/healthz'];
  let iteration = 0;

  const timer = setInterval(() => {
    iteration += 1;
    const route = routes[Math.floor(Math.random() * routes.length)];
    const durationMs = Math.floor(Math.random() * 500);
    let status = 200;

    // Emit inside an active span so the log record carries TraceId/SpanId.
    const span = tracer.startSpan('handle_request');
    context.with(trace.setSpan(context.active(), span), () => {
      const roll = Math.random();
      if (roll < 0.1) {
        status = 500;
        logger.emit({
          severityNumber: SeverityNumber.ERROR,
          severityText: 'ERROR',
          body: 'request failed',
          attributes: {
            'http.method': 'GET',
            'http.route': route,
            'http.status_code': status,
            'duration_ms': durationMs,
            'error': 'upstream timeout',
          },
        });
      } else if (durationMs > 350) {
        logger.emit({
          severityNumber: SeverityNumber.WARN,
          severityText: 'WARN',
          body: 'slow request',
          attributes: {
            'http.method': 'GET',
            'http.route': route,
            'http.status_code': status,
            'duration_ms': durationMs,
          },
        });
      } else {
        logger.emit({
          severityNumber: SeverityNumber.INFO,
          severityText: 'INFO',
          body: 'request handled',
          attributes: {
            'http.method': 'GET',
            'http.route': route,
            'http.status_code': status,
            'duration_ms': durationMs,
          },
        });
      }
    });
    span.end();

    if (iteration % 10 === 0) console.log(`  [${iteration}] logs sent`);
  }, 2000);

  const shutdown = async () => {
    clearInterval(timer);
    console.log('\nShutting down...');
    try {
      await loggerProvider.shutdown();
      await tracerProvider.shutdown();
    } catch (err) {
      console.error('shutdown error:', err);
    }
    console.log('Done!');
    process.exit(0);
  };
  process.on('SIGINT', shutdown);
  process.on('SIGTERM', shutdown);
}

main();
